From 6d7bfedd497fcc536e7b1505c3be8293c70ac7a0 Mon Sep 17 00:00:00 2001
From: Sam Protsenko <semen.protsenko@linaro.org>
Date: Thu, 12 Jul 2018 19:20:10 +0300
Subject: [PATCH] ARM: dts: am335x-boneblack: Re-mux USER/BOOT button pin

USER/BOOT button is soldered to LCD_DATA2 pin. In current device tree
this pin is muxed in LCD_DATA2 mode, which is needed for HDMI (in
am335x-boneblack-common.dtsi file, nxp_hdmi_bonelt_pins node).

For educational purposes we need this pin to be muxed in GPIO mode. To
make it possible let's do next:
 - disable HDMI chip that uses that pin (TDA19988)
 - add pinmux entry for that pin (mux as GPIO input, as we are going to
   use it for interrupt catching)
 - use pinmux entry in "soc" node, so that pinctrl subsystem actually
   remuxes that pin in GPIO mode

When booting with this dtb file, we can request GPIO2_8 line and read
its input value (or request interrupt on that line change). GPIO number
is easy to calculate:

    gpio_num = gpio_lines_per_bank * bank_num + gpio_num

On AM335x we have 32 GPIO lines in one GPIO bank. First GPIO bank is
gpio0. So gpio_num for GPIO2_8 line is:

    gpio_num = 32 * 2 + 8 = 72

Use that number in your kernel module to request GPIO line. Now when you
press USER/BOOT button, you can see the change on GPIO line in your
module, either by reading GPIO value of that line, or by configuring the
interrupt handler (ISR) for that GPIO line.

The last thing worth mentioning: this line is pulled-up with external
resistor on your board (you can see it on publically available
schematic). That's why we don't pull-up this line with internal resistor
in device tree. Also, this means that you will see:
 - "1" when button is not pressed
 - "0" when button is pressed
And if you want to request interrupt for that line, be sure to configure
it for "falling" change (from "1" to "0"), exactly for reason stated
above.

Signed-off-by: Sam Protsenko <semen.protsenko@linaro.org>
---
 arch/arm/boot/dts/am335x-boneblack.dts | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/arch/arm/boot/dts/am335x-boneblack.dts b/arch/arm/boot/dts/am335x-boneblack.dts
index d154d3133c16..cb1b642964f3 100644
--- a/arch/arm/boot/dts/am335x-boneblack.dts
+++ b/arch/arm/boot/dts/am335x-boneblack.dts
@@ -26,3 +26,22 @@
 		opp-supported-hw = <0x06 0x0100>;
 	};
 };
+
+&tda19988 {
+	status = "disabled";
+};
+
+&am33xx_pinmux {
+	user_button_pins: user_button_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8a8, PIN_INPUT | MUX_MODE7)		/* lcd_data2.gpio2_8 */
+		>;
+	};
+};
+
+/ {
+	soc {
+		pinctrl-names = "default";
+		pinctrl-0 = <&user_button_pins>;
+	};
+};
-- 
2.18.0

